src/daemon/control.rs

```diff
@@ -886,6 +886,159 @@ pub fn handle_rpc_messages(
                     outpoints,
                 )?)?;
             }
+            RpcMessageIn::GetSpendTx(outpoints, destinations, feerate_vb, response_tx) => {
+                log::trace!("Got 'getspendtx' request from RPC thread");
+                let revaultd = revaultd.read().unwrap();
+                let xpub_ctx = revaultd.xpub_ctx();
+                let db_file = &revaultd.db_file();
+
+                // Reconstruct the DepositTxin s from the outpoints and the vaults informations
+                let mut txins = Vec::with_capacity(outpoints.len());
+                // If we need a change output, use the highest derivation index of the vaults
+                // spent. This avoids leaking a new address needlessly while not introducing
+                // disrepancy between our indexes.
+                let mut change_index = ChildNumber::from(0);
+                for outpoint in outpoints.iter() {
+                    match db_vault_by_deposit(db_file, outpoint)? {
+                        Some(vault) => match vault.status {
+                            VaultStatus::Active => {
+                                let deposit_descriptor =
+                                    revaultd.deposit_descriptor.derive(vault.derivation_index);
+                                if vault.derivation_index > change_index {
+                                    change_index = vault.derivation_index;
+                                }
+                                txins.push((
+                                    DepositTxIn::new(
+                                        *outpoint,
+                                        DepositTxOut::new(
+                                            vault.amount.as_sat(),
+                                            &deposit_descriptor,
+                                            xpub_ctx,
+                                        ),
+                                    ),
+                                    vault.derivation_index,
+                                ));
+                            }
+                            status => {
+                                response_tx.send(Err(RpcControlError::InvalidStatus((
+                                    status,
+                                    VaultStatus::Active,
+                                ))))?;
+                                break;
+                            }
+                        },
+                        None => {
+                            response_tx.send(Err(RpcControlError::UnknownOutpoint(*outpoint)))?;
+                            break;
+                        }
+                    }
```

Comment:
I prefer if this in a `fn(vault) -> Result<DepositTxin, RpcControlError>`

Reply:
I do too but when i tried it got clumsy. Will retry..

Reply:
Ok, i actually tried to move too much to a function, with your solution it's cleaner :)
```diff
diff --git a/src/daemon/control.rs b/src/daemon/control.rs
index 96f7d01..21e61ba 100644
--- a/src/daemon/control.rs
+++ b/src/daemon/control.rs
@@ -455,6 +455,37 @@ fn share_unvault_signatures(
     send_sig_msg(&mut transport, txid, sigs.clone())
 }
 
+// Derive the DepositTxIn (to be referenced in an Unvault transaction) out of an existing
+// deposit outpoint.
+fn deposit_txin_from_outpoint(
+    revaultd: &RevaultD,
+    outpoint: &OutPoint,
+) -> Result<Result<(DepositTxIn, ChildNumber), RpcControlError>, ControlError> {
+    match db_vault_by_deposit(&revaultd.db_file(), outpoint)? {
+        Some(vault) => match vault.status {
+            VaultStatus::Active => {
+                let deposit_descriptor = revaultd.deposit_descriptor.derive(vault.derivation_index);
+                Ok(Ok((
+                    DepositTxIn::new(
+                        *outpoint,
+                        DepositTxOut::new(
+                            vault.amount.as_sat(),
+                            &deposit_descriptor,
+                            revaultd.xpub_ctx(),
+                        ),
+                    ),
+                    vault.derivation_index,
+                )))
+            }
+            status => Ok(Err(RpcControlError::InvalidStatus((
+                status,
+                VaultStatus::Active,
+            )))),
+        },
+        None => Ok(Err(RpcControlError::UnknownOutpoint(*outpoint))),
+    }
+}
+
 /// Handle events incoming from the JSONRPC interface.
 pub fn handle_rpc_messages(
     revaultd: Arc<RwLock<RevaultD>>,
@@ -899,39 +930,18 @@ pub fn handle_rpc_messages(
                 // disrepancy between our indexes.
                 let mut change_index = ChildNumber::from(0);
                 for outpoint in outpoints.iter() {
-                    match db_vault_by_deposit(db_file, outpoint)? {
-                        Some(vault) => match vault.status {
-                            VaultStatus::Active => {
-                                let deposit_descriptor =
-                                    revaultd.deposit_descriptor.derive(vault.derivation_index);
-                                if vault.derivation_index > change_index {
-                                    change_index = vault.derivation_index;
-                                }
-                                txins.push((
-                                    DepositTxIn::new(
-                                        *outpoint,
-                                        DepositTxOut::new(
-                                            vault.amount.as_sat(),
-                                            &deposit_descriptor,
-                                            xpub_ctx,
-                                        ),
-                                    ),
-                                    vault.derivation_index,
-                                ));
+                    let (deposit_txin, deriv_index) =
+                        match deposit_txin_from_outpoint(&revaultd, &outpoint)? {
+                            Ok(res) => res,
+                            Err(e) => {
+                                response_tx.send(Err(e))?;
+                                continue;
                             }
-                            status => {
-                                response_tx.send(Err(RpcControlError::InvalidStatus((
-                                    status,
-                                    VaultStatus::Active,
-                                ))))?;
-                                break;
-                            }
-                        },
-                        None => {
-                            response_tx.send(Err(RpcControlError::UnknownOutpoint(*outpoint)))?;
-                            break;
-                        }
+                        };
+                    if deriv_index > change_index {
+                        change_index = deriv_index;
                     }
+                    txins.push((deposit_txin, deriv_index));
                 }
                 if txins.len() != outpoints.len() {
                     // There was an error with one of the vaults, error is already sent.
diff --git a/tests/servers/coordinatord b/tests/servers/coordinatord
--- a/tests/servers/coordinatord
+++ b/tests/servers/coordinatord
@@ -1 +1 @@
-Subproject commit aefa59617be6dce12583edd3e772ed89ee550d07
+Subproject commit aefa59617be6dce12583edd3e772ed89ee550d07-dirty
```